### PR TITLE
Add another argument for the JERRY_STATIC_ASSERT with the description of the assert statement.

### PR DESCRIPTION
**Pull request by robertsipka**
The release.linux build fails with enabled ALL_IN_ONE option.

### CODE REVIEW

**Comment by akiss77** [comment]
I got confused by the title and the comment of this PR when I took a look at the patch. All I saw was that static asserts were extended with an additional message-like argument. I'm still not sure why that's needed. What kind of typedefs get redefined?

Other than that, the patch seems to be a fair step-by-step rewrite of all static asserts, that's good. (If that's really needed.)

Finally, please make sure that all touched files have the copyright year(s) updated in their headers. (At least some of them contain 2015.)

File: jerry-core/ecma/base/ecma-alloc.c [hunk 1]
```diff
@@ -20,16 +20,24 @@
...
+JERRY_STATIC_ASSERT (sizeof (ecma_property_t) <= sizeof (uint64_t),
+                     size_of_ecma_property_t_must_be_less_than_or_equal_to_8_bytes);
 
-JERRY_STATIC_ASSERT (sizeof (ecma_object_t) <= sizeof (uint64_t));
-JERRY_STATIC_ASSERT (ECMA_OBJECT_OBJ_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE);
-JERRY_STATIC_ASSERT (ECMA_OBJECT_LEX_ENV_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE);
+JERRY_STATIC_ASSERT (sizeof (ecma_object_t) <= sizeof (uint64_t),
+                     size_of_ecma_object_t_must_be_less_than_or_equal_to_8_bytes);
+JERRY_STATIC_ASSERT (ECMA_OBJECT_OBJ_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE,
+                     ECMA_OBJECT_OBJ_TYPE_SIZE_must_be_less_than_or_equal_to_64_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-alloc.c]: `_to_8_bytes` (!)

File: jerry-core/ecma/base/ecma-alloc.c [hunk 2]
```diff
@@ -20,16 +20,24 @@
...
+JERRY_STATIC_ASSERT (ECMA_OBJECT_LEX_ENV_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE,
+                     ECMA_OBJECT_LEX_ENV_TYPE_SIZE_must_be_less_than_or_equal_to_64_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-alloc.c]: ditto

File: jerry-core/ecma/base/ecma-alloc.c [hunk 3]
```diff
@@ -20,16 +20,24 @@
...
 
-JERRY_STATIC_ASSERT (sizeof (ecma_collection_header_t) == sizeof (uint64_t));
-JERRY_STATIC_ASSERT (sizeof (ecma_collection_chunk_t) == sizeof (uint64_t));
-JERRY_STATIC_ASSERT (sizeof (ecma_string_t) == sizeof (uint64_t));
-JERRY_STATIC_ASSERT (sizeof (ecma_getter_setter_pointers_t) <= sizeof (uint64_t));
+JERRY_STATIC_ASSERT (sizeof (ecma_collection_header_t) == sizeof (uint64_t),
+                     ecma_collection_header_t_must_be_less_than_or_equal_to_8_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-alloc.c]: `size_of_` ?

File: jerry-core/ecma/base/ecma-alloc.c [hunk 4]
```diff
@@ -20,16 +20,24 @@
...
+JERRY_STATIC_ASSERT (sizeof (ecma_collection_chunk_t) == sizeof (uint64_t),
+                     ecma_collection_chunk_t_must_be_less_than_or_equal_to_8_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-alloc.c]: ditto

File: jerry-core/ecma/base/ecma-alloc.c [hunk 5]
```diff
@@ -20,16 +20,24 @@
...
+JERRY_STATIC_ASSERT (sizeof (ecma_string_t) == sizeof (uint64_t),
+                     ecma_string_t_must_be_less_than_or_equal_to_8_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-alloc.c]: ditto?

File: jerry-core/ecma/base/ecma-alloc.c [hunk 6]
```diff
@@ -20,16 +20,24 @@
...
+JERRY_STATIC_ASSERT (sizeof (ecma_getter_setter_pointers_t) <= sizeof (uint64_t),
+                     ecma_getter_setter_pointers_t_must_be_less_than_or_equal_to_8_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-alloc.c]: ditto

File: jerry-core/ecma/base/ecma-helpers-external-pointers.c [hunk 7]
```diff
@@ -119,7 +120,8 @@ ecma_get_external_pointer_value (ecma_object_t *obj_p, /**< object to get proper
     return false;
   }
 
-  JERRY_STATIC_ASSERT (sizeof (uint32_t) <= sizeof (prop_p->u.internal_property.value));
+  JERRY_STATIC_ASSERT (sizeof (uint32_t) <= sizeof (prop_p->u.internal_property.value),
+                       size_of_internal_property_value_must_be_greater_than_or_equal_to_4_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-external-pointers.c]: This raises the question why this static assert is needed at all? It's a verbatim copy of the static assert above, which makes it useless.

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 8]
```diff
@@ -42,7 +42,8 @@
  * The length should be representable with int32_t.
  */
 JERRY_STATIC_ASSERT ((uint32_t) ((int32_t) ECMA_STRING_MAX_CONCATENATION_LENGTH) ==
-                     ECMA_STRING_MAX_CONCATENATION_LENGTH);
+                     ECMA_STRING_MAX_CONCATENATION_LENGTH,
+                     the_length_should_be_representable_with_int32_t);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-string.c]: `ECMA_STRING_MAX_CONCATENATION_LENGTH_should_be_`...
(And why that double type cast? It surely must have a tricky reason but it's not obvious to me.)

File: jerry-core/ecma/base/ecma-helpers-value.c [hunk 9]
```diff
@@ -28,7 +28,8 @@
 #include "jrt-bit-fields.h"
 #include "vm-defines.h"
 
-JERRY_STATIC_ASSERT (sizeof (ecma_value_t) * JERRY_BITSINBYTE >= ECMA_VALUE_SIZE);
+JERRY_STATIC_ASSERT (sizeof (ecma_value_t) * JERRY_BITSINBYTE >= ECMA_VALUE_SIZE,
+                     size_of_ecma_value_t_multiply_by_JERRY_BITSINBYTE_must_be_greater_or_equal_to_ECMA_VALUE_SIZE);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-value.c]: `bits_in_ecma_value_t_must_be_greater_or_equal_to_ECMA_VALUE_SIZE` ?

File: jerry-core/ecma/base/ecma-helpers.c [hunk 10]
```diff
@@ -444,7 +444,8 @@ ecma_create_internal_property (ecma_object_t *object_p, /**< the object */
   ECMA_SET_POINTER (new_property_p->next_property_p, list_head_p);
   ecma_set_property_list (object_p, new_property_p);
 
-  JERRY_STATIC_ASSERT (ECMA_INTERNAL_PROPERTY__COUNT <= (1ull << ECMA_PROPERTY_INTERNAL_PROPERTY_TYPE_WIDTH));
+  JERRY_STATIC_ASSERT (ECMA_INTERNAL_PROPERTY__COUNT <= (1ull << ECMA_PROPERTY_INTERNAL_PROPERTY_TYPE_WIDTH),
+                       ECMA_INTERNAL_PROPERTY__COUNT_must_be_less_than_or_equal_to_1ull_left_shifting_with_5_bits);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers.c]: `ECMA_PROPERTY_INTERNAL_PROPERTY_TYPE_WIDTH_bits_must_be_able_to_represent_ECMA_INTERNAL_PROPERTY__COUNT_values` ?

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.c [hunk 11]
```diff
@@ -108,7 +108,8 @@ ecma_builtin_global_object_print (ecma_value_t this_arg __attr_unused___, /**< t
       }
       else
       {
-        JERRY_STATIC_ASSERT (sizeof (code_point) == 2);
+        JERRY_STATIC_ASSERT (sizeof (code_point) == 2,
+                             size_of_code_point_must_be_equal_to_2);
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.c]: `_bytes`

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 12]
```diff
@@ -314,7 +314,8 @@ ecma_builtin_try_to_instantiate_property (ecma_object_t *object_p, /**< object *
                                                                ECMA_INTERNAL_PROPERTY_BUILT_IN_ROUTINE_DESC);
       uint64_t builtin_routine_desc = desc_prop_p->u.internal_property.value;
 
-      JERRY_STATIC_ASSERT (sizeof (uint8_t) * JERRY_BITSINBYTE == ECMA_BUILTIN_ROUTINE_ID_LENGTH_VALUE_WIDTH);
+      JERRY_STATIC_ASSERT (sizeof (uint8_t) * JERRY_BITSINBYTE == ECMA_BUILTIN_ROUTINE_ID_LENGTH_VALUE_WIDTH,
+                           size_of_ECMA_BUILTIN_ROUTINE_ID_LENGTH_VALUE_WIDTH_must_be_equal_to_8_bytes);
```
Comment (akiss77) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: `ECMA_BUILTIN_ROUTINE_ID_LENGTH_VALUE_WIDTH_must_be_64`

File: jerry-core/mem/mem-heap.c [hunk 13]
```diff
@@ -106,7 +106,8 @@ typedef enum __attr_packed___
 /**
  * Chunk size should satisfy the required alignment value
  */
-JERRY_STATIC_ASSERT (MEM_HEAP_CHUNK_SIZE % MEM_ALIGNMENT == 0);
+JERRY_STATIC_ASSERT (MEM_HEAP_CHUNK_SIZE % MEM_ALIGNMENT == 0,
+                     MEM_HEAP_CHUNK_SIZE_modulo_MEM_ALIGNMENT_must_be_equal_to_zero);
```
Comment (akiss77) [inline jerry-core/mem/mem-heap.c]: `MEM_HEAP_CHUNK_SIZE_must_be_multiple_of_MEM_ALIGNMENT`

File: jerry-core/mem/mem-heap.c [hunk 14]
```diff
@@ -160,7 +161,8 @@ typedef size_t mem_heap_bitmap_storage_item_t;
 /**
  * Overall number of bitmap bits is multiple of number of bits in a bitmap storage item
  */
-JERRY_STATIC_ASSERT (MEM_HEAP_BITMAP_BITS % MEM_HEAP_BITMAP_BITS_IN_STORAGE_ITEM == 0);
+JERRY_STATIC_ASSERT (MEM_HEAP_BITMAP_BITS % MEM_HEAP_BITMAP_BITS_IN_STORAGE_ITEM == 0,
+                     MEM_HEAP_BITMAP_BITS_modulo_MEM_HEAP_BITMAP_BITS_IN_STORAGE_ITEM_must_be_equal_to_zero);
```
Comment (akiss77) [inline jerry-core/mem/mem-heap.c]: as above

File: jerry-core/mem/mem-heap.c [hunk 15]
```diff
@@ -329,10 +332,14 @@ mem_heap_mark_chunk_allocated (size_t chunk_index, /**< index of the heap chunk
...
-  JERRY_STATIC_ASSERT ((uintptr_t) mem_heap.area % MEM_HEAP_CHUNK_SIZE == 0);
-  JERRY_STATIC_ASSERT (MEM_HEAP_AREA_SIZE % MEM_HEAP_CHUNK_SIZE == 0);
+  JERRY_STATIC_ASSERT ((MEM_HEAP_CHUNK_SIZE & (MEM_HEAP_CHUNK_SIZE - 1u)) == 0,
+                       MEM_HEAP_CHUNK_SIZE_bitand_MEM_HEAP_CHUNK_SIZE_minus_1u_must_be_equal_to_zero);
+  JERRY_STATIC_ASSERT ((uintptr_t) mem_heap.area % MEM_ALIGNMENT == 0,
+                       mem_heap_area_modulo_MEM_ALIGNMENT_must_be_equal_to_zero);
+  JERRY_STATIC_ASSERT ((uintptr_t) mem_heap.area % MEM_HEAP_CHUNK_SIZE == 0,
+                        mem_heap_area_modulo_MEM_HEAP_CHUNK_SIZE_must_be_equal_to_zero);
+  JERRY_STATIC_ASSERT (MEM_HEAP_AREA_SIZE % MEM_HEAP_CHUNK_SIZE == 0,
+                       MEM_HEAP_AREA_SIZE_modulo_MEM_HEAP_CHUNK_SIZE_must_be_equal_to_zero);
```
Comment (akiss77) [inline jerry-core/mem/mem-heap.c]: `_must_be_multiple_of_` (3x)

File: jerry-core/mem/mem-heap.c [hunk 16]
```diff
@@ -329,10 +332,14 @@ mem_heap_mark_chunk_allocated (size_t chunk_index, /**< index of the heap chunk
 void
 mem_heap_init (void)
 {
-  JERRY_STATIC_ASSERT ((MEM_HEAP_CHUNK_SIZE & (MEM_HEAP_CHUNK_SIZE - 1u)) == 0);
-  JERRY_STATIC_ASSERT ((uintptr_t) mem_heap.area % MEM_ALIGNMENT == 0);
-  JERRY_STATIC_ASSERT ((uintptr_t) mem_heap.area % MEM_HEAP_CHUNK_SIZE == 0);
-  JERRY_STATIC_ASSERT (MEM_HEAP_AREA_SIZE % MEM_HEAP_CHUNK_SIZE == 0);
+  JERRY_STATIC_ASSERT ((MEM_HEAP_CHUNK_SIZE & (MEM_HEAP_CHUNK_SIZE - 1u)) == 0,
+                       MEM_HEAP_CHUNK_SIZE_bitand_MEM_HEAP_CHUNK_SIZE_minus_1u_must_be_equal_to_zero);
```
Comment (akiss77) [inline jerry-core/mem/mem-heap.c]: `MEM_HEAP_CHUNK_SIZE_must_be_power_of_2`

File: jerry-core/mem/mem-heap.c [hunk 17]
```diff
@@ -757,8 +764,10 @@ mem_heap_free_block (void *ptr) /**< pointer to beginning of data space of the b
 void *
 mem_heap_get_chunked_block_start (void *ptr) /**< pointer into a block */
 {
-  JERRY_STATIC_ASSERT ((MEM_HEAP_CHUNK_SIZE & (MEM_HEAP_CHUNK_SIZE - 1u)) == 0);
-  JERRY_STATIC_ASSERT (((uintptr_t) mem_heap.area % MEM_HEAP_CHUNK_SIZE) == 0);
+  JERRY_STATIC_ASSERT ((MEM_HEAP_CHUNK_SIZE & (MEM_HEAP_CHUNK_SIZE - 1u)) == 0,
+                       MEM_HEAP_CHUNK_SIZE_bitand_MEM_HEAP_CHUNK_SIZE_minus_1u_must_be_equal_to_zero);
+  JERRY_STATIC_ASSERT (((uintptr_t) mem_heap.area % MEM_HEAP_CHUNK_SIZE) == 0,
+                       mem_heap_area_modulo_MEM_HEAP_CHUNK_SIZE_must_be_equal_to_zero);
```
Comment (akiss77) [inline jerry-core/mem/mem-heap.c]: These are duplicates too. Strange.

File: jerry-core/mem/mem-poolman.c [hunk 18]
```diff
@@ -567,9 +568,12 @@ mem_pools_alloc_longpath (void)
   mem_free_chunks_number += MEM_POOL_CHUNKS_NUMBER;
 #endif /* !JERRY_NDEBUG */
 
-  JERRY_STATIC_ASSERT (MEM_POOL_CHUNK_SIZE % MEM_ALIGNMENT == 0);
-  JERRY_STATIC_ASSERT (sizeof (mem_pool_chunk_t) == MEM_POOL_CHUNK_SIZE);
-  JERRY_STATIC_ASSERT (sizeof (mem_pool_chunk_index_t) <= MEM_POOL_CHUNK_SIZE);
+  JERRY_STATIC_ASSERT (MEM_POOL_CHUNK_SIZE % MEM_ALIGNMENT == 0,
+                       MEM_POOL_CHUNK_SIZE_modulo_MEM_ALIGNMENT_must_be_equal_to_zero);
```
Comment (akiss77) [inline jerry-core/mem/mem-poolman.c]: multiple of

File: jerry-core/mem/mem-poolman.c [hunk 19]
```diff
@@ -567,9 +568,12 @@ mem_pools_alloc_longpath (void)
...
+  JERRY_STATIC_ASSERT (sizeof (mem_pool_chunk_t) == MEM_POOL_CHUNK_SIZE,
+                       size_of_mem_pool_chunk_t_must_be_equal_to_MEM_POOL_CHUNK_SIZE);
```
Comment (akiss77) [inline jerry-core/mem/mem-poolman.c]: duplicate?...

File: jerry-core/rcs/rcs-records.c [hunk 20]
```diff
@@ -255,7 +255,8 @@ rcs_record_get_pointer (rcs_record_t *rec_p, /**< record */
 rcs_record_type_t
 rcs_record_get_type (rcs_record_t *rec_p) /**< record */
 {
-  JERRY_STATIC_ASSERT (sizeof (rcs_record_type_t) * JERRY_BITSINBYTE >= RCS_HEADER_TYPE_WIDTH);
+  JERRY_STATIC_ASSERT (sizeof (rcs_record_type_t) * JERRY_BITSINBYTE >= RCS_HEADER_TYPE_WIDTH,
+                       size_of_rcs_record_type_t_multiply_by_JERRY_BITSINBYTE_must_be_greater_than_or_equal_to_4u);
```
Comment (akiss77) [inline jerry-core/rcs/rcs-records.c]: `bits_in_rcs_record_type_t_must_be_greater_than_or_equal_to_RCS_HEADER_TYPE_WIDTH`

**Comment by zherczeg** [comment]
The generic direction is good, but I also agree with Akos that some names need to be a bit better.

**Comment by robertsipka** [comment]
Thanks for the comments, I've updated the patch with the suggested names.

File: jerry-core/ecma/base/ecma-alloc.c [hunk 21]
```diff
@@ -20,16 +21,24 @@
...
+JERRY_STATIC_ASSERT (sizeof (ecma_property_t) <= sizeof (uint64_t),
+                     size_of_ecma_property_t_must_be_less_than_or_equal_to_8_bytes);
 
-JERRY_STATIC_ASSERT (sizeof (ecma_object_t) <= sizeof (uint64_t));
-JERRY_STATIC_ASSERT (ECMA_OBJECT_OBJ_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE);
-JERRY_STATIC_ASSERT (ECMA_OBJECT_LEX_ENV_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE);
+JERRY_STATIC_ASSERT (sizeof (ecma_object_t) <= sizeof (uint64_t),
+                     size_of_ecma_object_t_must_be_less_than_or_equal_to_8_bytes);
+JERRY_STATIC_ASSERT (ECMA_OBJECT_OBJ_TYPE_SIZE <= sizeof (uint64_t) * JERRY_BITSINBYTE,
+                     ECMA_OBJECT_OBJ_TYPE_SIZE_must_be_less_than_or_equal_to_bits_in_uint64_t);
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-alloc.c]: ECMA_OBJECT_OBJ_TYPE_SIZE_must_be_less_than_or_equal_to_64_bits ? A bit shorter.

**Comment by zherczeg** [comment]
Patch looks much better.

**Comment by robertsipka** [comment]
Thanks, I've updated, and I've also removed the duplicated static asserts too.

File: jerry-core/mem/mem-heap.c [hunk 22]
```diff
@@ -1,4 +1,5 @@
 /* Copyright 2014-2015 Samsung Electronics Co., Ltd.
```
Comment (LaszloLango) [inline jerry-core/mem/mem-heap.c]: `/* Copyright 2014-2016 Samsung Electronics Co., Ltd.`

File: jerry-core/mem/mem-poolman.c [hunk 23]
```diff
@@ -1,4 +1,5 @@
 /* Copyright 2014-2015 Samsung Electronics Co., Ltd.
```
Comment (LaszloLango) [inline jerry-core/mem/mem-poolman.c]: ditto

**Comment by akiss77** [comment]
Patch looks good, thanks. But still, could you please provide a description of the error it fixes? At least for future reference.

**Comment by robertsipka** [comment]
@akiss77 : The JERRY_STATIC_ASSERT defines a variable, where the name contains the __LINE__ of the assert. So, in case of all-in-one build, it's possible that there are at least two variables which have the same name (redefinition), because an assert occurs in the same line in another file.

**Comment by akiss77** [comment]
@robertsipka thanks, LGTM (FWIW)

**Comment by zherczeg** [comment]
LGTM

**Comment by LaszloLango** [comment]
LGTM

**Pull request merged**